contracts/consensus/Coconsensus.sol

```diff
@@ -0,0 +1,203 @@
+    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) public domainSeparators;
+
+
+    /* Special Functions */
+
+    /**
+     * @notice Setup function does the initialization of all the mosaic
+     *         contracts on the auxiliary chain.
+     *
+     * @dev This function can be called only once.
+     */
+    function setup() public {
+
+        require(
+            selfMetachainId == bytes32(0),
+            "Coconsensus contract is already initialized."
+        );
+
+        originMetachainId = genesisOriginMetachainId;
+
+        selfMetachainId = genesisSelfMetachainId;
+
+        relativeSelfDynasty = uint256(0);
+
+        bytes32 currentMetachainId = genesisMetachainIds[SENTINEL_METACHAIN_ID];
+
+        // Loop through the genesis metachainId link list.
+        while (currentMetachainId != SENTINEL_METACHAIN_ID) {
+
+            // Setup protocore contract for the given metachain id.
+            setupProtocore(currentMetachainId);
+
+            // Setup observer contract for the given metachain id.
+            setupObserver(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* Private Functions */
+
+    /**
+     * @notice Do the initial setup of protocore contract and initialize the
+     *         storage of coconsensus contract.
+     *
+     * @param _metachainId Metachain id.
+     *
+     * \pre Protocore contract must exist for given metachain id.
+     *
+     * \post Adds newly setup protocore's address into protocores storage variable.
+     * \post Adds newly setup protocore's domain separator into domainSeparators.
+     *       storage variable.
+     * \post Adds a new Block into blockchain storage variable.
+     * \post Updates blockTips storage variable with the latest finalized
+     *       checkpoint's block number of the newly setup protocore.
+     */
+    function setupProtocore(bytes32 _metachainId) private {
+
+        // Get the protocore contract address from the genesis storage.
+        address protocoreAddress = genesisProtocores[_metachainId];
+
+        require(
+            protocoreAddress != address(0),
+            "Protocore address must not be null."
+        );
+
+        // Setup protocore.
+        ProtocoreI protocore = ProtocoreI(protocoreAddress);
+        ( bytes32 blockHash, uint256 blockNumber ) = protocore.setup();
+
+        // Store the protocore address in protocores mapping.
+        protocores[_metachainId] = protocore;
+
+        // Get the domain separator and store it in domainSeparators mapping.
+        domainSeparators[_metachainId] = protocore.domainSeparator();
+
+        // Store the block informations in blockchains mapping.
+        blockchains[_metachainId][blockNumber] = Block(
+            blockHash,
+            CheckpointCommitStatus.Finalized,
```

Comment:
OriginObserver will have some initial state root anchored. It may be for the same block number. 
We can take it as a point for discussion. Keeping this code as it is.

cc: @jayeshbairagi